content/en/blog/guestbook-application-with-fission-golang.md

```diff
@@ -0,0 +1,310 @@
+
+```bash
+$ fission httptrigger create --method GET \
+    --url "/guestbook/messages/{id:[0-9]+}" --function restapi-get
+```
+
+### Function
+
+To handle a REST API request, normally a function need to extract resource from the request URL.
+In Fission, you can get the resource value from request header directly as we described in [Request Payload](#request-payload).
+
+That means you can get value from header with key `X-Fission-Params-Id` if URL is `/guestbook/messages/{id}`.
+
+In Golang, you can use `CanonicalMIMEHeaderKey` to transform letter case.
+
+* [rest-api/api.go](https://github.com/fission/fission-restapi-sample/blob/f011b08cf08412250c295a534f71fb381fd30e4d/rest-api/api.go#L83-L100)
+
+```go
+import (
+    "net/textproto"
+)
+
+const (
+    URL_PARAMS_PREFIX = "X-Fission-Params"
+)
+
+func GetPathValue(r *http.Request, param string) string {
+    // transform text case
+    // For example: 'id' -> 'Id', 'fooBAR' -> 'Foobar', 'foo-BAR' -> 'Foo-Bar'.
+    param = textproto.CanonicalMIMEHeaderKey(param)
+
+    // generate header key for accessing request header value
+    key := fmt.Sprintf("%s-%s", URL_PARAMS_PREFIX, param)
+
+    // get header value
+    return r.Header.Get(key)
+}
+```
+
+When Golang server receive HTTP requests, it dispatches `entrypoint function` and passes whole `response writer` and `request` object to the function.
+
+```go
+func MessageGetHandler(w http.ResponseWriter, r *http.Request) {
+    ...
+}
+```
+
+In this way, you can access `query string` and `message body` as usual.
+
+* [rest-api/api.go](https://github.com/fission/fission-restapi-sample/blob/4398eb195aeb59523101aa68a62782d91f86d85a/rest-api/api.go#L65-L67)
+
+```go
+func GetQueryString(r *http.Request, query string) string {
+    return r.URL.Query().Get(query)
+}
+```
+
+* [rest-api/api.go](https://github.com/fission/fission-restapi-sample/blob/4398eb195aeb59523101aa68a62782d91f86d85a/rest-api/api.go#L103-L110)
+
+```go
+body, err := ioutil.ReadAll(r.Body)
+if err != nil {
+    err = errors.Wrap(err, "Error reading request body")
+    log.Println(err.Error())
+    w.WriteHeader(http.StatusBadRequest)
+    w.Write([]byte(err.Error()))
+    return
+}
+```
+
+## CockroachDB Setup
+
+In guestbook application we need to store messages in a database.
+Since fission is build on top of kubernetes, it's very easy to achieve this by creating Kubernetes `Service(svc)` and `Deployment`.
+For this sample application, we are using [CockroachDB](https://www.cockroachlabs.com/docs/cockroachcloud/quickstart.html).
+
+> We also have a post on using [PostgreSQL with Fission](../how-to-use-postgresql-database-with-fission-functions/) that you might want to read too.
+
+![Access 3rd service](/images/how-to-develop-a-serverless-application-with-fission/access-3rd-service.svg)
+
+Add `cockroachdb` helm repo
+
+``` bash
+helm repo add cockroachdb https://charts.cockroachdb.com/
+helm repo update
+```
+
+Install cockroachdb using helm
```

Comment:
Before this you can add the content or link to cockroachdb.yaml file.

Reply:
Done